src/backend/dx12/src/device.rs

```diff
@@ -160,31 +163,66 @@ impl Device {
             })
     }
 
-    fn patch_spirv_resources(ast: &mut spirv::Ast<hlsl::Target>) -> Result<(), d::ShaderError> {
+    fn patch_spirv_resources(
+        ast: &mut spirv::Ast<hlsl::Target>,
+        layout: Option<&n::PipelineLayout>,
+    ) -> Result<(), d::ShaderError> {
         // Patch descriptor sets due to the splitting of descriptor heaps into
         // SrvCbvUav and sampler heap. Each set will have a new location to match
         // the layout of the root signatures.
+        let space_offset = match layout {
+            Some(layout) if !layout.root_constants.is_empty() => 1,
+            _ => 0,
+        };
+
         let shader_resources = ast.get_shader_resources().map_err(gen_query_error)?;
         for image in &shader_resources.separate_images {
             let set = ast.get_decoration(image.id, spirv::Decoration::DescriptorSet).map_err(gen_query_error)?;
-            ast.set_decoration(image.id, spirv::Decoration::DescriptorSet, 2*set)
+            ast.set_decoration(image.id, spirv::Decoration::DescriptorSet, space_offset + 2*set)
                .map_err(gen_unexpected_error)?;
         }
 
         for sampler in &shader_resources.separate_samplers {
             let set = ast.get_decoration(sampler.id, spirv::Decoration::DescriptorSet).map_err(gen_query_error)?;
-            ast.set_decoration(sampler.id, spirv::Decoration::DescriptorSet, 2*set+1)
+            ast.set_decoration(sampler.id, spirv::Decoration::DescriptorSet, space_offset + 2*set+1)
                .map_err(gen_unexpected_error)?;
         }
 
         Ok(())
     }
 
-    fn translate_spirv(ast: &mut spirv::Ast<hlsl::Target>, shader_model: hlsl::ShaderModel) -> Result<String, d::ShaderError> {
+    fn translate_spirv(
+        ast: &mut spirv::Ast<hlsl::Target>,
+        shader_model: hlsl::ShaderModel,
+        layout: &n::PipelineLayout,
+        stage: pso::Stage,
+    ) -> Result<String, d::ShaderError> {
         let mut compile_options = hlsl::CompilerOptions::default();
         compile_options.shader_model = shader_model;
         compile_options.vertex.invert_y = true;
 
+        let stage_flag = match stage {
```

Comment:
we should probably implement `From<pso::Stage> for pso::ShaderStageFlags`, if it isn't already